src/ocean/util/container/btree/BTreeMap.d

```diff
@@ -0,0 +1,664 @@
+
+    CPU complexity of the operations (n is number of the elements, t is degree,
+    h is the tree height):
+
+        - Searching: O(th) = O(t * log_t(n))
+        - Inserting: O(th) = O(t * log_t(n))
+        - Deleting:  O(th) = O(t * log_t(n))
+
+    Complexity of the memory access (to read/write the node from/in memory):
+
+        - Searching O(h) = O(log_t(n))
+        - Inserting O(h) = O(log_t(n))
+        - Deleting: O(h) = O(log_t(n))
+
+    References:
+        - https://en.wikipedia.org/wiki/B-tree
+        - Thomas H. Cormen et al. "Introduction to Algorithms, 3rd edition"
+
+    Params:
+        KeyType = type of the key.
+        TreeElementType = type of the element to store.
+        tree_degree = degree of the tree (refer to the documentation)
+
+******************************************************************************/
+
+struct BTreeMap(TreeKeyType, TreeElementType, int tree_degree)
+{
+    import ocean.util.container.btree.Implementation;
+
+    static assert (tree_degree > 0);
+
+    /**************************************************************************
+
+        Constructor.
+
+        Params:
+            allocator = memory manager to use to allocate/deallocate memory
+
+    ***************************************************************************/
+
+    package void init (IMemManager allocator = mallocMemManager)
+    {
+        this.impl.init(allocator);
+    }
+
+    /**************************************************************************
+
+        Inserts the element in the tree. This is passing the copy of the
+        element into the tree, so it's not necessary to manually create copy of
+        the element.  Note that for the reference types, this will just copy
+        the reference.
+
+        Params:
+            el = element to insert.
+
+        Returns:
+            true if the element did not exist and it was inserted, false otherwise
+
+    ***************************************************************************/
+
+    public bool insert (KeyType key, ElementType el)
+    {
+        return this.impl.insert(key, el);
+    }
+
+    /******************************************************************************
+
+        Deletes the element from the BTreeMap. The element is passed by reference,
+        since we'll just use it for the comparison with the value stored in the tree.
+
+        Params:
+            element = element to delete.
+
+        Returns:
+            true if the element was found and deleted, false otherwise.
+
+     ******************************************************************************/
+
+    public bool remove (KeyType key)
+    {
+        return this.impl.remove(key);
+    }
+
+    /******************************************************************************
+
+        Finds the highest value in the tree.
+
+        Params:
+            el = out parameter to be filled with the copy of the highest element
```

Comment:
Yes, this is more useful for the set, it's not very useful now. I'll remove it, since the ordering is done by keys and this would return the item with the highest key.